### DISCUSSION TITLE
U8g2: SSD1306 on ESP32 "bugging" after some time

### DISCUSSION THREAD

**Discussion by ezcGman** [olikraus/u8g2#1647] [Q&A] [answered]
Hey there,

I'm currently developing a mod for a pretty famous LED control software (WLED) which (in my specific case) runs on an ESP32 on an analog LED dimmer board.

The hardware facts first:
- Using u8g2 on a "QuinLED-ESP32", which is a reimplementation of the famous WEMOS D1 Mini 32. Difference is that this one has an WROOM-32D with 8MB flash size: https://quinled.info/quinled-esp32/
  - In the pictures below you will see an ethernet hat ontop of the ESP32: The issues I describe below happened also without this ethernet hat, so it's unrelated.
- Display is an 128x64 SSD1306 two color display from AliExpress: https://www.aliexpress.com/item/1005001621832273.html
- Using (software) SPI to control it, pins: 0 (CS), 32 (DC),  33 (RST), 16 (MOSI), 15 (CLK)
- The ESP and OLED sit on that mentioned LED dimmer board, but that shouldn't matter; the IOs are routed directly from the ESP to the pin socket the OLED sits on.

So my issue is that after some time, the OLED "bugs out" / "flickers" / whatever u wanna call it. I made two videos showing what I mean:
- How it should be: https://imgur.com/a/puY0JDB
- Bugging out after some time: https://imgur.com/a/eyIRXyp

This certainly looks like a buffer mismatch / overflow or paging issue. So I've tried all three different buffer modes already (1, 2 F), but same result. The time after this happens also is totally random. Sometimes on boot, sometimes after a few seconds, sometimes it takes minutes, sometimes it seems it's not happening at all. With the same routine updating the display.
I've also tried various bus clock speeds, from not setting it at all / leave default, to 1MHz or 40MHz (software SPI max on ESP32): Same outcome as described above.

My code is a little bigger, but roughly explained it circles through three "pages", staying on one page for 10 seconds, each showing different information. One page is showing the current PWM on the LED outs (what you can see in the videos above), so constant redrawing. The other two are just "static" text, so drawn once and waiting 10 seconds / no redraw.
The bugging out also sometimes happens when drawing the "static text" pages, so not only on that PWM "page".

Some snippets:
Initialization of the display:
```
// As mentioned, I tried all three buffer modes
oledDisplay = (U8G2 *) new U8G2_SSD1306_128X64_NONAME_F_4W_SW_SPI(U8G2_R0, oledSpiClk, oledSpiData, oledSpiCs, oledSpiDc, oledSpiRst);
if (oledDisplay == nullptr) {
  oledEnabled = oledInitDone = false;
  return;
}

oledDisplay->begin();
oledDisplay->setBusClock(40 * 1000 * 1000); // As mentioned, I tried with/without this and different values
oledDisplay->setContrast(10);
oledDisplay->setPowerSave(0);
oledDisplay->setFont(u8g2_font_6x10_tf);
```
After this, I draw an XBM...
```
oledDisplay->clearBuffer();
oledDisplay->drawXBMP(0, 16, 128, 36, quinLedLogo);
oledDisplay->sendBuffer();
oledLogoDrawn = millis();
```
... then (within the loop()) I wait until three seconds have passed since then and then start drawing the "pages", but with 150ms of waiting in the loop() before I attempt to redraw:
```
if (isOledReady() && millis() - oledLogoDrawn > 3000 && millis() - oledLastTimeUpdated > 150) {
  // Performing some additional checks here, but no calls to the OLED
  oledShowNextPage();
  oledLastTimeUpdated = millis();
}
```
Inside the oledShowNextPage() it looks roughly like this:
```
oledDisplay->clearBuffer();
oledDisplay->setFont(u8g2_font_chroma48medium8_8r);
oledDisplay->drawStr(0, 8, serverDescription);
oledDisplay->drawHLine(0, 13, 127);
oledDisplay->setFont(u8g2_font_6x10_tf);

switch (oledCurrentPage) {
  // LED Outputs
  case 1:
    // Doing things like drawStr(), drawRFrame() & drawBox()
    break;

  case 2:
    // Only a few drawStr() calls
    break;

  case 3:
    // Also only a few drawStr() calls
    break;
}

oledDisplay->sendBuffer();
```
As mentioned, I've also tried the buffer modes 1 and 2, so in the above examples wherever there is a call to `oledDisplay->clearBuffer();`, replace it with `oledDisplay->firstPage(); do {`. And wherever there is a call to `oledDisplay->sendBuffer();`, replace it with `} while (oledDisplay->nextPage());`.

I'm currently clueless what else I can do ad hope somebody has seen this before or has a lead what could cause this.
Thanks already and if any more info is needed, I'm happy to provide that!

Greetings,

Andy!

**Comment by olikraus** [comment, marked as answer]
If another call to oledDisplay->begin(); would recover your display, then it seems that the OLED looses some configuration values.
Maybe you can call oledDisplay->begin(); every 10 seconds or so.

I could imagine, that the PWM signal, which drives your LEDs, have a bad influence on the OLED display.

**Reply by ezcGman** [reply]
Hey Oli,

thanks for your reply!

Another call to begin() indeed recovers the display, forgot to mention that. I just didn't add that to the routine/loop, as I didn't know if that's fine to do.
Gonna try this then!

Interesting that the OLED "forgets" the config because of PWM, if that's really the case 🤔

I actually try to drive the OLED with a few jumper wires, farer away from the board, maybe that changes something / won't make it bug out.

Thanks again!

**Reply by olikraus** [reply]
> Interesting that the OLED "forgets" the config because of PWM,

It could be also a power supply issue, if too many LEDs are driven at the same time. In such a case the power supply voltage my drop, which might cause the OLED to forget some data.

A capacitor close to the OLED power supply may also help.

**Reply by ezcGman** [reply]
I haven't been driving any LEDs actually when this happens, none are connected to the board.
Board is powered by a PSU with 150W 24V, so that is more than fine ;)
I think there is a cap near the OLED actually but I need to double check.

**Reply by ezcGman** [reply]
Sorry for the late response! I've been testing with
a) Not having any PWM going on / LEDs turned off / MOSFETs not switching and the display seems to "survive"; I'm not getting any issues.
b) Calling `begin()` periodically: This also works and makes the display "come back" / work again.
So thanks for your fast replies here, much appreciated!

And I realized I haven't said this yet: Absolutely amazing job with this library! Awesome work!

Greetings,

Andy!
